Support symbol retargeting for function pointers.

333fred opened the pull request:
@AlekseyTs @dotnet/roslyn-compiler for review.

333fred commented:
Note that there was no PROTOTYPE comment associated with this, it had already been converted to a bullet in https://github.com/dotnet/roslyn/issues/39865

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler for a review.

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -845,6 +921,41 @@ public void CheckNamedTypes(NamedTypeSymbol a, NamedTypeSymbol b)
+                checkModifiers(aParam.RefCustomModifiers, bParam.RefCustomModifiers);
+            }
+
+            void checkModifiers(ImmutableArray<CustomModifier> aModifiers, ImmutableArray<CustomModifier> bModifiers)
```
Believe this can be `static`

333fred replied:
No, I call `CheckSymbols`. That's an instance method.

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -892,6 +1003,11 @@ protected override void CheckSymbolsInternal(Symbol a, Symbol b)
+            if (a is FunctionPointerTypeSymbol && b is FunctionPointerTypeSymbol)
+            {
+                return;
+            }
```
Why do we need to skip the check here for `FunctionPointerTypeSymbol`?

333fred replied:
Because function pointers don't have underlying symbols, it's a meaningless check.

333fred replied:
If we had pointer symbol tests in this file, we'd have to bail out for them as well.

jaredpar replied:
Consider adding a comment here

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,82 @@ class C1<T>
+            var sourceAssembly = (SourceAssemblySymbol)sourceModule.ContainingAssembly;
+            var sourceNamespace = sourceModule.GlobalNamespace;
+
+            var retargetingAssembly = new RetargetingAssemblySymbol(sourceAssembly, isLinked: false);
```
>var retargetingAssembly = new RetargetingAssemblySymbol(sourceAssembly, isLinked: false); [](start = 12, length = 89)

Let's not do this. Let's get retargeting assembly created by the system and than assert our expectations.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,82 @@ class C1<T>
+            var retargetingModule = retargetingAssembly.Modules[0];
+            var retargetingNamespace = retargetingModule.GlobalNamespace;
+
+            RetargetingSymbolChecker.CheckSymbols(sourceNamespace.GetMember<NamedTypeSymbol>("C"), retargetingNamespace.GetMember<NamedTypeSymbol>("C"));
```
>RetargetingSymbolChecker.CheckSymbols( [](start = 12, length = 38)

I have no idea what this this thing is doing and why do we need to call it and fix it up.

AlekseyTs replied:
If the goal is to test that the function pointer symbol is not getting modified in the retargeting process, I recommend to use approach taken by the test below. Only don't change mscorelib version. Change the version of some other reference, could be any other library, even custom one. 

---
In reply to: [414901658](https://github.com/dotnet/roslyn/pull/43625#discussion_r414901658) [](ancestors = 414901658)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -845,6 +921,41 @@ public void CheckNamedTypes(NamedTypeSymbol a, NamedTypeSymbol b)
             CheckSymbols(a.TypeParameters, b.TypeParameters, true);
         }
 
+        public void CheckFunctionPointerTypes(FunctionPointerTypeSymbol a, FunctionPointerTypeSymbol b)
```
>CheckFunctionPointerTypes [](start = 20, length = 25)

It looks like this method is not used anywhere. Remove?

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,82 @@ class C1<T>
+            Assert.Same(c2CorLib, param_c2.Type.ContainingAssembly);
+
+            Assert.NotSame(c2CorLib, ((CSharpCustomModifier)param_c1.RefCustomModifiers.Single()).ModifierSymbol.ContainingAssembly);
+            Assert.Same(c2CorLib, ((CSharpCustomModifier)param_c2.RefCustomModifiers.Single()).ModifierSymbol.ContainingAssembly);
```
>Assert.Same(c2CorLib, ((CSharpCustomModifier)param_c2.RefCustomModifiers.Single()).ModifierSymbol.ContainingAssembly); [](start = 12, length = 118)

It would be good to test:
- retargeting of ref modifiers on the signature
- retargeting of modifiers on return types and parameter types. However, the way to get them there would be through inheritance during overriding and I am not sure if that piece is implemented yet.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.cs:
```diff
@@ -729,6 +729,56 @@ public PointerTypeSymbol Retarget(PointerTypeSymbol type)
+                    }
+                }
+
+                if (symbolModified)
```
>symbolModified [](start = 20, length = 14)

It doesn't look like added tests actually verify that this clever way of tracking if anything is changed actually works properly. I believe we are testing when everything is changed and nothing is changed. I think it is worth adding a test when  a change happens in one specific place and go though all of them one by one:
- ref modifier on return
- modifier on return
- return type
- ref modifier on first param (at least two params in total) 
- modifier on first param type (at least two params in total) 
- type on first param (at least two params in total) 
- ref modifier on second param
- modifier on second param type
- type on second param

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
@AlekseyTs addressed feedback. I totally rewrote the tests in RetargetingTests.cs (and that's the only change made), so I'd suggest not looking at the diff between iteration 1 and 2 and just reviewing RetargetingTests.cs from scratch as the diff is a bit messy.

---
In reply to: [619274690](https://github.com/dotnet/roslyn/pull/43625#issuecomment-619274690) [](ancestors = 619274690)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,250 @@ class C1<T>
+            {
+                var retargetedSource = @"
+using System.Reflection;
+[assembly:AssemblyVersionAttribute(""{0}"")]]
```
This feels more complicated than it could be. CreateCompilation takes an AssemblyIdentity instance (at least there is an overload that does), this how usually we specify version in tests rather than changing/parameterizing the source.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,250 @@ class C1<T>
+                var retargeted2Ref = retargeted2.ToMetadataReference();
+
+
+                var consistent = CreateCompilation("public class C {}", assemblyName: "Con");
```
>var [](start = 16, length = 3)

Extra empty line above

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,250 @@ class C1<T>
+
+            assert(returnConsistent.typeConsistent,
+                   ptrOriginal.Signature.ReturnType.ContainingAssembly,
+                   ptrRetargeted.Signature.ReturnType.ContainingAssembly);
```
>ptrRetargeted.Signature.ReturnType [](start = 19, length = 34)

Consider strengthening the test and asserting that none of the types (including modifiers) in the retargeted signature are error types.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,250 @@ class C1<T>
+                   getModifierAssemblySymbol(param2Original.TypeWithAnnotations.CustomModifiers),
+                   getModifierAssemblySymbol(param2Retargeted.TypeWithAnnotations.CustomModifiers));
+
+            static MethodSymbol getMethodSymbol(CSharpCompilation compilation)
```
>static [](start = 12, length = 6)

It feels like we are lacking an assert that the function pointer type stayed the same if all types involved expected to be "consistent".

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +750,250 @@ class C1<T>
         }
+
+        [Fact]
+        public void FunctionPointerRetargeting_FullyConsistent()
```
>public [](start = 8, length = 6)

```[CompilerTrait(CompilerFeature....)]```? And for other new tests

333fred replied:
I'll do this in a separate PR. I haven't added a CompilerFeature flag nor have I done this for any existing tests. I'd want to either do everything or nothing.

---
In reply to: [418119854](https://github.com/dotnet/roslyn/pull/43625#discussion_r418119854) [](ancestors = 418119854)

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented on src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:
```diff
@@ -750,6 +751,262 @@ class C1<T>
+                var retargeted2 = CreateCompilation(retargetedIdentity.WithVersion(new Version(2, 0, 0, 0)), new[] { retargetedSource }, references: standardReference);
+                var retargeted2Ref = retargeted2.ToMetadataReference();
+
+                var consistent = CreateCompilation("public class C {}", assemblyName: "Con", targetFramework: TargetFramework.Standard);
```
Because I had to add an explicit StandardReference above (because that CreateCompilation overload does not add default references), I'm being explicit in the others rather than relying on any change to CreateCompilation to change for all of the calls.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [621949028](https://github.com/dotnet/roslyn/pull/43625#issuecomment-621949028) [](ancestors = 621949028)

333fred commented:
@dotnet/roslyn-compiler for a second review.

333fred commented:
Rebased on top of the latest master-merge
